### PR TITLE
regrtest computes statistics

### PR DESCRIPTION
**Pull request by vstinner**
* Rewrite TestResult as a dataclass with a new State class.
* Add test.support.TestStats class and Regrtest.stats_dict attribute.
* libregrtest.runtest functions now modify a TestResult instance in-place.
* libregrtest summary lists the number of run tests and skipped tests, and denied resources.

### CODE REVIEW

**Comment by vstinner** [comment]
Example on Linux:

```
$ ./python -u -m test -j0 -u all -r
(...)

== Tests result: SUCCESS ==

450 tests OK.

12 tests skipped:
    test.test_asyncio.test_windows_events
    test.test_asyncio.test_windows_utils test_devpoll test_ioctl
    test_kqueue test_launcher test_startfile test_winconsoleio
    test_winreg test_winsound test_wmi test_zipfile64

Total cases: run=42,205 skipped=1,027
Total duration: 13 min 1 sec
Tests result: SUCCESS (12 skipped; 1 resource denied)
```

The "resource denied" test is ``test_zipfile64``: it requires ``extralargefile`` resource.

```
0:03:00 load avg: 16.56 [157/462] test_zipfile64 skipped (resource denied) -- (...)
```

The ``extralargefile`` resource is **not** enabled by ``-u all``, extract of ``libregrtest/cmdline.py``:

```py
# Other resources excluded from --use=all:
#
# - extralagefile (ex: test_zipfile64): really too slow to be enabled
#   "by default"
# - tzdata: while needed to validate fully test_datetime, it makes
#   test_datetime too slow (15-20 min on some buildbots) and so is disabled by
#   default (see bpo-30822).
RESOURCE_NAMES = ALL_RESOURCES + ('extralargefile', 'tzdata')
```


The 12 skipped tests are:

```
test_kqueue skipped -- test works only on BSD
test_winreg skipped -- No module named 'winreg'
test_zipfile64 skipped -- test requires loads of disk-space bytes and a long time to run
test_startfile skipped -- object <module 'os' from '/home/vstinner/python/main/Lib/os.py'> has no attribute 'startfile'
test_devpoll skipped -- test works only on Solaris OS family
test_ioctl skipped -- Unable to open /dev/tty
test_winconsoleio skipped -- test only relevant on win32
test_wmi skipped -- No module named '_wmi'
test.test_asyncio.test_windows_utils skipped -- Windows only
test_winsound skipped -- No module named 'winsound'
test.test_asyncio.test_windows_events skipped -- Windows only
test_launcher skipped -- test only applies to Windows
```

**Comment by vstinner** [comment]
I'm not sure how to name the added statistics:

```
Total cases: run=42,205 skipped=1,027
```

The term "test" is misleading since it can mean different things: a test **file**, a test **case** (class), or a test **method**.

Would it be better to write ``Total cases`` here? Something like:

```
450 tests OK.

Total tests: run=42,205 skipped=1,027
```

Is it clear that 450 is the number of test **files** and 42,205 the number of test **methods**?

By the way, Python 3.13 now has test **packages** like test_asyncio and test_concurrent_futures. A package itself is made of multiple test files. The term "test" alone is even more confusing.

The doctest module counts the number of ``>>> `` test *statements*. A single function can have multiple test *statements*.

**Comment by vstinner** [comment]
Results on GitHub Action jobs.

## Ubuntu

```
449 tests OK.

10 slowest tests:
- test_gdb: 3 min 13 sec
- test.test_multiprocessing_spawn.test_processes: 1 min 52 sec
- test.test_multiprocessing_forkserver.test_processes: 1 min 17 sec
- test.test_concurrent_futures.test_wait: 1 min 14 sec
- test_subprocess: 1 min 6 sec
- test_socket: 1 min 2 sec
- test_cppext: 1 min 2 sec
- test_venv: 57.6 sec
- test.test_multiprocessing_spawn.test_misc: 56.4 sec
- test_signal: 54.8 sec

13 tests skipped:
    test.test_asyncio.test_windows_events
    test.test_asyncio.test_windows_utils test_devpoll test_ioctl
    test_kqueue test_launcher test_peg_generator test_startfile
    test_winconsoleio test_winreg test_winsound test_wmi
    test_zipfile64

Total cases: run=42,126 skipped=1,131
Total duration: 11 min 55 sec
Tests result: SUCCESS (13 skipped; 2 resource denied)
```

## macOS

```

440 tests OK.

10 slowest tests:
- test.test_multiprocessing_spawn.test_processes: 2 min 39 sec
- test_signal: 2 min 27 sec
- test_cppext: 2 min 19 sec
- test.test_multiprocessing_forkserver.test_processes: 1 min 49 sec
- test_ssl: 1 min 42 sec
- test_tarfile: 1 min 40 sec
- test_largefile: 1 min 32 sec
- test_statistics: 1 min 20 sec
- test.test_concurrent_futures.test_wait: 1 min 16 sec
- test_zipfile: 1 min 13 sec

22 tests skipped:
    test.test_asyncio.test_windows_events
    test.test_asyncio.test_windows_utils
    test.test_multiprocessing_fork.test_manager
    test.test_multiprocessing_fork.test_misc
    test.test_multiprocessing_fork.test_processes
    test.test_multiprocessing_fork.test_threads test_devpoll
    test_epoll test_gdb test_ioctl test_launcher test_peg_generator
    test_perf_profiler test_perfmaps test_startfile test_tkinter
    test_ttk test_winconsoleio test_winreg test_winsound test_wmi
    test_zipfile64

Total cases: run=40,314 skipped=1,431
Total duration: 16 min 7 sec
Tests result: SUCCESS (22 skipped; 4 resource denied)
```

**Comment by vstinner** [comment]
regrtest fails to count statistics of test_xml_etree_c(). Its ``test_main()`` function returns None:

```
Warning -- test_xml_etree_c test runner returned None: <function test_main at 0x7f96d5326a50>
```

This one can be fixed later.

File: Lib/test/libregrtest/runtest.py [hunk 1]
```diff
@@ -10,115 +12,102 @@
...
+        elif state == State.INTERRUPTED:
+            return f"{self.test_name} interrupted"
+        elif state == State.CHILD_ERROR:
+            return f"{self.test_name} crashed"
+        elif state == State.DID_NOT_RUN:
+            return f"{self.test_name} ran no tests"
+        elif state == State.TIMEOUT:
+            return f"{self.test_name} timed out ({format_duration(self.duration)})"
+        else:
+            raise ValueError("unknown result state: {state!r}")
```
Comment (AlexWaygood) [inline Lib/test/libregrtest/runtest.py]: Consider using pattern-matching here -- I find it much more elegant for this kind of usage:

```suggestion
        match self.state:
            case State.PASSED:
                return f"{self.test_name} passed"
            case State.FAILED:
                return self._format_failed()
            case State.SKIPPED:
                return f"{self.test_name} skipped"
            case State.UNCAUGHT_EXC:
                return f"{self.test_name} failed (uncaught exception)"
            case State.REFLEAK:
                return f"{self.test_name} failed (reference leak)"
            case State.ENV_CHANGED:
                return f"{self.test_name} failed (env changed)"
            case State.RESOURCE_DENIED:
                return f"{self.test_name} skipped (resource denied)"
            case State.INTERRUPTED:
                return f"{self.test_name} interrupted"
            case State.CHILD_ERROR:
                return f"{self.test_name} crashed"
            case State.DID_NOT_RUN:
                return f"{self.test_name} ran no tests"
            case State.TIMEOUT:
                return f"{self.test_name} timed out ({format_duration(self.duration)})"
            case _:
                raise ValueError("unknown result state: {state!r}")
```

File: Lib/test/libregrtest/main.py [hunk 2]
```diff
@@ -115,32 +115,36 @@ def get_executed(self):
...
-        test_name = result.name
+        test_name = result.test_name
 
-        if not isinstance(result, (ChildError, Interrupted)) and not rerun:
-            self.test_times.append((result.duration_sec, test_name))
+        if not(result.state == State.CHILD_ERROR or result.state == State.INTERRUPTED) and not rerun:
+            self.test_times.append((result.duration, test_name))
 
-        if isinstance(result, Passed):
+        state = result.state
```
Comment (AlexWaygood) [inline Lib/test/libregrtest/main.py]: Pattern-matching could improve the readability of this block, too (the diff is relative to your PR branch):

```diff
diff --git a/Lib/test/libregrtest/main.py b/Lib/test/libregrtest/main.py
index b0de7f879c..0616778adc 100644
--- a/Lib/test/libregrtest/main.py
+++ b/Lib/test/libregrtest/main.py
@@ -120,26 +120,26 @@ def accumulate_result(self, result, rerun=False):
         if not(result.state == State.CHILD_ERROR or result.state == State.INTERRUPTED) and not rerun:
             self.test_times.append((result.duration, test_name))

-        state = result.state
-        if state == State.PASSED:
-            self.good.append(test_name)
-        elif state == State.ENV_CHANGED:
-            self.environment_changed.append(test_name)
-        elif result.is_failed(self.ns.fail_env_changed):
-            if not rerun:
-                self.bad.append(test_name)
-                self.need_rerun.append(result)
-        elif state == State.SKIPPED:
-            self.skipped.append(test_name)
-        elif state == State.RESOURCE_DENIED:
-            self.skipped.append(test_name)
-            self.resource_denieds.append(test_name)
-        elif state == State.INTERRUPTED:
-            self.interrupted = True
-        elif state == State.DID_NOT_RUN:
-            self.run_no_tests.append(test_name)
-        else:
-            raise ValueError(f"invalid test state: {state!r}")
+        match result.state:
+            case State.PASSED:
+                self.good.append(test_name)
+            case State.ENV_CHANGED:
+                self.environment_changed.append(test_name)
+            case state if result.is_failed(self.ns.fail_env_changed):
+                if not rerun:
+                    self.bad.append(test_name)
+                    self.need_rerun.append(result)
+            case State.SKIPPED:
+                self.skipped.append(test_name)
+            case State.RESOURCE_DENIED:
+                self.skipped.append(test_name)
+                self.resource_denieds.append(test_name)
+            case State.INTERRUPTED:
+                self.interrupted = True
+            case State.DID_NOT_RUN:
+                self.run_no_tests.append(test_name)
+            case _:
+                raise ValueError(f"invalid test state: {state!r}")
```

File: Lib/test/libregrtest/runtest.py [hunk 3]
```diff
@@ -10,115 +12,118 @@
...
-        self.errors = errors
-        self.failures = failures
+    @staticmethod
+    def is_failed(state):
+        return state in (
+            State.FAILED,
+            State.UNCAUGHT_EXC,
+            State.REFLEAK,
+            State.MULTIPROCESSING_ERROR,
+            State.TIMEOUT)
```
Comment (AlexWaygood) [inline Lib/test/libregrtest/runtest.py]: nit (feel free to ignore if you disagree)

```suggestion
        return state in {
            State.FAILED,
            State.UNCAUGHT_EXC,
            State.REFLEAK,
            State.MULTIPROCESSING_ERROR,
            State.TIMEOUT}
```

File: Lib/test/libregrtest/runtest.py [hunk 4]
```diff
@@ -10,115 +12,118 @@
...
+        # Consider that the duration is meaningless for these cases.
+        # For example, if a whole test file is skipped, its duration
+        # is unlikely to be the duration of executing its tests,
+        # but just the duration to execute code which skips the test.
+        return state not in (
+            State.SKIPPED,
+            State.RESOURCE_DENIED,
+            State.INTERRUPTED,
+            State.MULTIPROCESSING_ERROR,
+            State.DID_NOT_RUN)
```
Comment (AlexWaygood) [inline Lib/test/libregrtest/runtest.py]:
```suggestion
        return state not in {
            State.SKIPPED,
            State.RESOURCE_DENIED,
            State.INTERRUPTED,
            State.MULTIPROCESSING_ERROR,
            State.DID_NOT_RUN}
```

File: Lib/test/libregrtest/runtest.py [hunk 5]
```diff
@@ -293,18 +292,44 @@ def _test_module(the_module):
...
+    elif isinstance(test_result, unittest.TestResult):
+        stats = TestStats.from_unittest(test_result)
+    elif isinstance(test_result, doctest.TestResults):
+        stats = TestStats.from_doctest(test_result)
+    elif test_result is None:
+        print_warning(f"{result.test_name} test runner returned None: {test_func}")
+        stats = None
+    else:
+        print_warning(f"Unknown test result type: {type(test_result)}")
+        stats = None
```
Comment (AlexWaygood) [inline Lib/test/libregrtest/runtest.py]:
```suggestion
    match test_result:
        case TestStats():
            stats = test_result
        case unittest.TestResult():
            stats = TestStats.from_unittest(test_result)
        case doctest.TestResults():
            stats = TestStats.from_doctest(test_result)
        case None:
            print_warning(f"{result.test_name} test runner returned None: {test_func}")
            stats = None
        case _:
            print_warning(f"Unknown test result type: {type(test_result)}")
            stats = None
```

File: Lib/test/libregrtest/runtest.py [hunk 6]
```diff
@@ -367,52 +383,66 @@ def _runtest_inner(
...
+        pass
+    elif support.environment_altered:
+        result.state = State.ENV_CHANGED
+    else:
+        result.state = State.PASSED
+    return result
 
 
-def cleanup_test_droppings(test_name: str, verbose: int) -> None:
+def cleanup_test_droppings(test_name: str, verbose: int):
```
Comment (AlexWaygood) [inline Lib/test/libregrtest/runtest.py]: What's the motivation for this change? The return annotation looks like it is still accurate

File: Lib/test/libregrtest/runtest.py [hunk 7]
```diff
@@ -194,7 +193,7 @@ def get_abs_module(ns: Namespace, test_name: str) -> str:
         return 'test.' + test_name
 
 
-def _runtest(ns: Namespace, test_name: str) -> TestResult:
+def _runtest(result: TestResult, ns: Namespace):
```
Comment (AlexWaygood) [inline Lib/test/libregrtest/runtest.py]: Most of the functions in this file appear to have type annotations; I'd prefer it if we could keep to the existing style here and have a return annotation for this function

File: Lib/test/libregrtest/runtest.py [hunk 8]
```diff
@@ -293,18 +292,44 @@ def _test_module(the_module):
...
-    support.run_unittest(tests)
+    return support.run_unittest(tests)
 
 
 def save_env(ns: Namespace, test_name: str):
     return saved_test_environment(test_name, ns.verbose, ns.quiet, pgo=ns.pgo)
 
 
-def _runtest_inner2(ns: Namespace, test_name: str) -> bool:
+def regrtest_runner(result, test_func, ns):
```
Comment (AlexWaygood) [inline Lib/test/libregrtest/runtest.py]:
```suggestion
def regrtest_runner(result, test_func, ns) -> None:
```

File: Lib/test/libregrtest/runtest.py [hunk 9]
```diff
@@ -352,12 +369,11 @@ def _runtest_inner2(ns: Namespace, test_name: str) -> bool:
...
 
-    return refleak
+    return result
 
 
-def _runtest_inner(
-    ns: Namespace, test_name: str, display_failure: bool = True
-) -> TestResult:
+def _runtest_inner(result: TestResult, ns: Namespace,
+                   display_failure: bool = True):
```
Comment (AlexWaygood) [inline Lib/test/libregrtest/runtest.py]: Could you add a return type annotation here, like there is in the function you're replacing? Most of the functions in this file have complete type annotations

**Comment by vstinner** [comment]
@AlexWaygood: Yeah, I was thinking about match/case, and I forgot. Sure thing, I updated my PR to use match/case and address your latest review.

Oops, I also took it as an opportunity to add some other clenaup/enhancements. It's hard to resisit in the middle of such large refactoring.

**Comment by vstinner** [comment]
Windows x64 GitHub Action job:

```
424 tests OK.

10 slowest tests:
- test.test_multiprocessing_spawn.test_processes: 4 min 32 sec
- test_tarfile: 2 min 55 sec
- test.test_multiprocessing_spawn.test_manager: 2 min 49 sec
- test_regrtest: 2 min 49 sec
- test_importlib: 2 min 29 sec
- test_socket: 2 min 7 sec
- test_compileall: 2 min 4 sec
- test_launcher: 2 min
- test.test_multiprocessing_spawn.test_misc: 1 min 57 sec
- test_threading: 1 min 52 sec

38 tests skipped:
    test.test_asyncio.test_unix_events
    test.test_multiprocessing_fork.test_manager
    test.test_multiprocessing_fork.test_misc
    test.test_multiprocessing_fork.test_processes
    test.test_multiprocessing_fork.test_threads
    test.test_multiprocessing_forkserver.test_manager
    test.test_multiprocessing_forkserver.test_misc
    test.test_multiprocessing_forkserver.test_processes
    test.test_multiprocessing_forkserver.test_threads test_curses
    test_dbm_gnu test_dbm_ndbm test_devpoll test_epoll test_fcntl
    test_fork1 test_gdb test_grp test_ioctl test_kqueue test_openpty
    test_peg_generator test_perf_profiler test_perfmaps test_poll
    test_posix test_pty test_pwd test_readline test_resource
    test_syslog test_threadsignals test_wait3 test_wait4
    test_xxlimited test_xxtestfuzz test_zipfile64 test_zoneinfo

Total cases: run=40,372 skipped=2,312
Total duration: 23 min 37 sec
Tests result: SUCCESS (38 skipped; 2 resource denied)
```

**Comment by vstinner** [comment]
I merged my PR. Thanks a lot @AlexWaygood for your reviews!

> test_netrc, test_pep646_syntax and test_xml_etree now return results
> in the test_main() function.

``test_xml_etree_c`` is the last test which is unable to report the number of tests: its ``test_main()`` function returns None.

**Comment by vstinner** [comment]
In general, I'm trying to keep libregrtest the same in all stable branches. It seems like 3.11 has an outdated libregrtest :-(
